### PR TITLE
Add update syntax and some reserved words

### LINKED ISSUE #3932: Add simple update syntax to perform update by specifying primary key
## Feature request
Add simple update syntax to perform update by specifying primary key, example:

```
update orders set state = 'finished' where id = 1000;
```
This is just the first step to support a series of DML SQL syntaxes, we pick the easiest task to do first.

**Is your feature request related to a problem? Please describe.**
Add more SQL syntax support

**Describe the solution you'd like**
Add syntax, parser, planner, since it's updated by primary key, it can be planned into a standard partial update easily, which has been implemented recently. 

Update: 
* full row update is used for now, after some reconsideration, write only partial update is incomplete to realize `update where xxx` syntax
* a read-write transaction style select then full row update approach is used currently, this won't guarantee transaction isolation, but still it's useful, use this with this limitation in mind, do not perform updates with other concurrent loads/updates.

### PR DESCRIPTION
**Pull request by decster**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #3932

## Problem Summary(Required) ：
This PR adds update syntax, so users can update primary key tables using SQL. It adds update-related syntax parser, analyzer, and planer. 
* Only non-key columns are allowed to be updated
* To prevent misuse user must specify a where clause to prevent a full table update.
* Full row update is used for now, because after some reconsideration, write only partial update is incomplete to support `update where xxx` syntax
* A read-write transaction style(select then full row update) approach is used currently, this won't guarantee transaction isolation(need table lock mechanism, will design&implement in future), but still it's useful, use this with this limitation in mind, do not perform updates with other concurrent loads/updates.

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 93 / 185 (50.27%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|0|36|00.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ExecPlan.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|2|25|08.00%|
|:large_blue_circle:|com/starrocks/analysis/DmlStmt.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|15|23|65.22%|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnAssignment.java|6|7|85.71%|
|:large_blue_circle:|com/starrocks/analysis/UpdateStmt.java|13|15|86.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|54|58|93.10%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                item = new SelectListItem(new SlotRef(tableName, col.getName()), col.getName());
+            }
+            selectList.addItem(item);
+        }
+
+        TableRelation tableRelation = new TableRelation(tableName);
+        tableRelation.setTable(table);
+        Scope tableScope = analyzeTable(tableRelation, tableName, table, session);
+
+        SelectRelation selectRelation =
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java:69]: Same as above, using QueryStatement

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                item = new SelectListItem(new CastExpr(col.getType(), assign.getExpr()), col.getName());
+            } else {
+                item = new SelectListItem(new SlotRef(tableName, col.getName()), col.getName());
+            }
+            selectList.addItem(item);
+        }
+
+        TableRelation tableRelation = new TableRelation(tableName);
+        tableRelation.setTable(table);
+        Scope tableScope = analyzeTable(tableRelation, tableName, table, session);
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java:67]: It is not necessary to call analyze for tableRelation separately, the analyze of query will perform the logic of table analyze

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/ColumnAssignment.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    public String getColumn() {
+        return column;
+    }
+
+    public Expr getExpr() {
+        return expr;
+    }
+
+    @Override
+    public String toSql() {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/ColumnAssignment.java:25]: toSql is no longer a required method for the new ParseNode, and will no longer rely on toSql for the "to sql" function in the future. If there is no need for tosql, this method may not be added.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 4]
```diff
@@ -14,14 +14,16 @@ singleStatement
...
-    | EXPLAIN? INSERT INTO qualifiedName (WITH LABEL lable=identifier)? columnAliases?
+    | EXPLAIN? INSERT INTO qualifiedName (WITH LABEL label=identifier)? columnAliases?
     (queryStatement | (VALUES expressionsWithDefault (',' expressionsWithDefault)*))    #insert
     | CREATE TABLE (IF NOT EXISTS)? qualifiedName
         ('(' identifier (',' identifier)* ')')? comment?
         partitionDesc?
         distributionDesc?
         properties?
         AS queryStatement                                                               #createTableAsSelect
+    | (EXPLAIN (LOGICAL | VERBOSE | COSTS)?)?
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:25]: Insert Statement does not support other types of explain, I think Update should be consistent?

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4 [hunk 5]
```diff
@@ -6,6 +6,7 @@ tokens {
     CONCAT
 }
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:8]: COMMIT，MERGE，ROLLBACK is keyword，but no reserve in mysql。I suggest we keep it consistent with mysql, non-reserved words need to be added in nonReserved in StarRocks.g4

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 4 above)
Comment (decster) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:25]: How about all support those explain types? If it is useful in select, it is also useful in "insert xxx select"

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/ColumnAssignment.java (hunk 3 above)
Comment (decster) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/ColumnAssignment.java:25]: ok

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 4 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:25]: Yes, just ensure that the insert and update logic are consistent.

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4 (hunk 5 above)
Comment (decster) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocksLex.g4:8]: ok

File: fe/fe-core/src/main/java/com/starrocks/analysis/UpdateStmt.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.util.List;
+
+public class UpdateStmt extends DmlStmt {
+    private final TableName tableName;
+    private final List<ColumnAssignment> assignments;
+    private final Expr wherePredicate;
+
+    private Table table;
+    private SelectRelation updateRelation;
```
[the remark this answers is not in the document]
Comment (decster) [reply fe/fe-core/src/main/java/com/starrocks/analysis/UpdateStmt.java]: ok

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java (hunk 1 above)
Comment (decster) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java:61]: ok

**Comment by decster** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 118 / 156 (75.64%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ExecPlan.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|2|25|08.00%|
|:large_blue_circle:|com/starrocks/analysis/DmlStmt.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|33|36|91.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|29|31|93.55%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|21|22|95.45%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/UpdateStmt.java|15|15|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnAssignment.java|6|6|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 133 / 156 (85.26%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|7|25|28.00%|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|33|36|91.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|29|31|93.55%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|22|22|100.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ExecPlan.java|8|8|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/UpdateStmt.java|15|15|100.00%|
|:large_blue_circle:|com/starrocks/analysis/DmlStmt.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnAssignment.java|6|6|100.00%|

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by decster** [comment]
run starrocks_fe_unittest

**Review by HangyuanLiu — APPROVED** [review]

**Comment by decster** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 133 / 156 (85.26%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|7|25|28.00%|
|:large_blue_circle:|com/starrocks/sql/UpdatePlanner.java|33|36|91.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/UpdateAnalyzer.java|29|31|93.55%|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|22|22|100.00%|
|:large_blue_circle:|com/starrocks/sql/StatementPlanner.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/sql/plan/ExecPlan.java|8|8|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/analysis/UpdateStmt.java|15|15|100.00%|
|:large_blue_circle:|com/starrocks/analysis/DmlStmt.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/ColumnAssignment.java|6|6|100.00%|

File: fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AnalyzeUpdateTest.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+    @AfterClass
+    public static void tearDown() {
+        File file = new File(runningDir);
+        file.delete();
+    }
+
+    @Test
+    public void testSingle() {
+        analyzeFail("update tjson set v_json = '' where v_int = 1",
```
Comment (chaoyli) [inline fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AnalyzeUpdateTest.java:33]: How about where v_int in (1, 2, 3)

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AnalyzeUpdateTest.java (hunk 7 above)
Comment (decster) [reply fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AnalyzeUpdateTest.java:33]: this test is used to verify update syntax only support the primary key table, so the kind of predicate is irrelevant.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    public static void analyze(UpdateStmt updateStmt, ConnectContext session) {
+        TableName tableName = updateStmt.getTableName();
+        MetaUtils.normalizationTableName(session, tableName);
+        MetaUtils.getStarRocks(session, tableName);
+        Table table = MetaUtils.getStarRocksTable(session, tableName);
+
+        if (!(table instanceof OlapTable && ((OlapTable) table).getKeysType() == KeysType.PRIMARY_KEYS)) {
+            throw unsupportedException("only support updating primary key table");
+        }
+        if (updateStmt.getWherePredicate() == null) {
```
[the remark this answers is not in the document]
Comment (decster) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/UpdateAnalyzer.java:37]: Yes, it's supported

**Review by chaoyli — APPROVED** [review]

**Pull request merged**